File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

using namespace std;
// 제대로 끝말잇기하는지 -> 이전의 끝 알파벳 저장
// 반복되진 않는지 -> set
vector<int> solution(int n, vector<string> words) {
    unordered_set<string> word;
    
    int turn = 1; char bef = words[0][0];
    for(int i = 0; i < words.size(); i++){
        if(words[i][0] != bef){ // 이어가지 못함 탈락
            return {i % n + 1, i/n + 1};
        }else if(word.count(words[i])){ // 중복됨
            return {i % n + 1, i/n + 1};
        } else { // 이어가기
            bef = words[i][words[i].size()-1];
            word.insert(words[i]);
        }
    }

    // 가장 먼저 탈락하는 사람의 번호, 자신의 몇 번째에 탈락하는지 => {번호, 차례}
    return {0, 0};
}

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*; 

class Solution {
    HashSet<String> wordStack = new HashSet<>(); 
    
    public int[] solution(int n, String[] words) {
        int[] answer = {};
        int failNum = -1; 
        
        int turn = 1; 
        char bef = words[0].charAt(0); // 직전에 나왔던 문자
        int id = 0;
        
        while(turn <= words.length / n + 1){
            
            for(int i = 0; i < n; i++){
                id = (turn-1) * n + i;
                if(id >= words.length) return new int[]{0, 0};
                
                // 직전에 나왔던 문자로 시작하는지
                // 같은거 반복되진 않았는지
                if(words[id].charAt(0) != bef || wordStack.contains(words[id])){
                    failNum = i; break; 
                }else{
                    bef = words[id].charAt(words[id].length()-1); 
                    wordStack.add(words[id]);
                }
               
            }
            if(failNum != -1) break;
            
            turn++;
        }

        

        // 가장 먼저 탈락하는 사람의 번호와 그 사람이 자신의 몇 번째 차례에 탈락하는지를 구해서 return
        return new int[]{failNum+1, turn};
    }
}